File: DashboardService/DashboardService/Controllers/TicketsController.cs
using DashboardService.Contracts.Tickets;
using Microsoft.AspNetCore.Mvc;

namespace DashboardService.Controllers;

[ApiController]
[Route("api/[controller]")]
public sealed class TicketsController(ITicketService ticketService) : ControllerBase
{
    [HttpPost]
    [ProducesResponseType(typeof(TicketDto), StatusCodes.Status201Created)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    public async Task<ActionResult<TicketDto>> Create([FromBody] CreateTicketRequest request, CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(request.Title) || string.IsNullOrWhiteSpace(request.Description))
            return ValidationProblem("Title and Description are required.");

        var created = await ticketService.CreateAsync(request, cancellationToken);
        return CreatedAtAction(nameof(GetById), new { id = created.Id }, created);
    }

    [HttpGet]
    [ProducesResponseType(typeof(IReadOnlyList<TicketDto>), StatusCodes.Status200OK)]
    public async Task<ActionResult<IReadOnlyList<TicketDto>>> GetAll(
        [FromQuery] TicketStatus? status,
        [FromQuery] TicketPriority? priority,
        CancellationToken cancellationToken)
    {
        var items = await ticketService.GetAllAsync(new TicketQuery(status, priority), cancellationToken);
        return Ok(items);
    }

    [HttpGet("{id:guid}")]
    [ProducesResponseType(typeof(TicketDto), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<ActionResult<TicketDto>> GetById([FromRoute] Guid id, CancellationToken cancellationToken)
    {
        var ticket = await ticketService.GetByIdAsync(id, cancellationToken);
        return ticket is null ? NotFound() : Ok(ticket);
    }

    [HttpPatch("{id:guid}/status")]
    [ProducesResponseType(typeof(TicketDto), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<ActionResult<TicketDto>> UpdateStatus(
        [FromRoute] Guid id,
        [FromBody] UpdateTicketStatusRequest request,
        CancellationToken cancellationToken)
    {
        var updated = await ticketService.UpdateStatusAsync(id, request, cancellationToken);
        return updated is null ? NotFound() : Ok(updated);
    }

    [HttpPatch("{id:guid}/assign")]
    [ProducesResponseType(typeof(TicketDto), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<ActionResult<TicketDto>> Assign(
        [FromRoute] Guid id,
        [FromBody] AssignTicketRequest request,
        CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(request.AssignedAgentId))
            return ValidationProblem("AssignedAgentId is required.");

        var updated = await ticketService.AssignAsync(id, request, cancellationToken);
        return updated is null ? NotFound() : Ok(updated);
    }
}

File: server/src/tickets/tickets.controller.ts
import {
  BadRequestException,
  Body,
  Controller,
  Get,
  NotFoundException,
  Param,
  ParseUUIDPipe,
  Patch,
  Post,
  Query,
} from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';
import { TicketsService } from './tickets.service';
import {
  ApiAssignTicket,
  ApiCreateTicket,
  ApiGetTicketById,
  ApiGetTickets,
  ApiUpdateTicketStatus,
} from './tickets.swagger';
import {
  AssignTicketRequest,
  CreateTicketRequest,
  TicketDto,
  TicketPriority,
  TicketQuery,
  TicketStatus,
  UpdateTicketStatusRequest,
} from './tickets.types';
import { isBlank, parseOptionalEnum } from '../common/utils/validation.utils';

@ApiTags('tickets')
@Controller('api/Tickets')
export class TicketsController {
  constructor(private readonly ticketsService: TicketsService) {}

  @Post()
  @ApiCreateTicket()
  async create(@Body() request: CreateTicketRequest): Promise<TicketDto> {
    if (isBlank(request.title) || isBlank(request.description)) {
      throw new BadRequestException('Title and Description are required.');
    }

    return await this.ticketsService.create(request);
  }

  @Get()
  @ApiGetTickets()
  async getAll(
    @Query('status') status?: TicketStatus,
    @Query('priority') priority?: TicketPriority,
  ): Promise<TicketDto[]> {
    const query: TicketQuery = {
      status,
      priority,
    };

    

    return await this.ticketsService.getAll(query);
  }

  @Get(':id')
  @ApiGetTicketById()
  async getById(@Param('id', new ParseUUIDPipe()) id: string): Promise<TicketDto> {
    const ticket = await this.ticketsService.getById(id);
    if (!ticket) {
      throw new NotFoundException('Ticket not found.');
    }

    return ticket;
  }

  @Patch(':id/status')
  @ApiUpdateTicketStatus()
  async updateStatus(
    @Param('id', new ParseUUIDPipe()) id: string,
    @Body() request: UpdateTicketStatusRequest,
  ): Promise<TicketDto> {
    const updated = await this.ticketsService.updateStatus(id, request);
    if (!updated) {
      throw new NotFoundException('Ticket not found.');
    }

    return updated;
  }

  @Patch(':id/assign')
  @ApiAssignTicket()
  async assign(
    @Param('id', new ParseUUIDPipe()) id: string,
    @Body() request: AssignTicketRequest,
  ): Promise<TicketDto> {
    if (isBlank(request.assignedAgentId)) {
      throw new BadRequestException('AssignedAgentId is required.');
    }

    const updated = await this.ticketsService.assign(id, request);
    if (!updated) {
      throw new NotFoundException('Ticket not found.');
    }

    return updated;
  }
}
